client/v2/client.go

```diff
@@ -415,49 +417,52 @@ func (uc *udpclient) Write(bp BatchPoints) error {
 		pointstring := p.pt.RoundedString(d) + "\n"
 
 		if b.Len()+len(pointstring) >= uc.payloadSize {
-			if _, err := uc.conn.Write(b.Bytes()); err != nil {
-				return err
+			if b.Len() > 0 {
+				if _, err := uc.conn.Write(b.Bytes()); err != nil {
+					return err
+				}
 			}
 			b.Reset()
 		}
 
 		if b.Len()+len(pointstring) <= uc.payloadSize {
-			if _, err := b.WriteString(pointstring); err != nil {
-				return err
+			b.WriteString(pointstring)
+			continue
+		}
+
+		if p.Time().IsZero() {
+			if len(pointstring) > MaxPointSize {
+				return ErrLargePoint
 			}
-		} else {
-			if p.Time().IsZero() {
-				if len(pointstring) > MaxPointSize {
-					return fmt.Errorf("point size is greater than %dB", MaxPointSize)
-				}
-				if _, err := b.WriteString(pointstring); err != nil {
-					return err
-				}
-			} else {
-				points, err := p.pt.SplitN(uc.payloadSize - 2) // -2 because of `+ "\n"`
-				if err != nil {
-					return err
-				}
-				for _, sp := range points {
-					pointstring := sp.RoundedString(d) + "\n"
-
-					if b.Len()+len(pointstring) >= uc.payloadSize {
-						if _, err := uc.conn.Write(b.Bytes()); err != nil {
-							return err
-						}
-						b.Reset()
-					}
+			b.WriteString(pointstring)
+			continue
+		}
+
+		points, err := p.pt.SplitN(uc.payloadSize - 2) // -2 because of `+ "\n"`
+		if err != nil {
+			return err
+		}
+		for _, sp := range points {
+			pointstring := sp.RoundedString(d) + "\n"
 
-					if _, err := b.WriteString(pointstring); err != nil {
+			if b.Len()+len(pointstring) >= uc.payloadSize {
+				if b.Len() > 0 {
```

Comment:
This should actually be included in the outer conditional, and the check should probably be a strict greater-than. If appending the current point creates the max payload size, that is okay since it is an inclusive limit.

I see it was being treated as an exclusive limit, but the naming doesn't match up with that.